Move collector to librustc_mir::monomorphize

MaikKlein opened the pull request:
cc https://github.com/rust-lang/rust/issues/44334 and https://github.com/rust-lang/rust/issues/45276

* I moved the collector to rustc_mir

*  I renamed `TransItem` to `MonoItem`. _(I still need to fix up comments and variable names)_

* I got rid of `common.rs` and `monomorphize.rs` from `librustc_trans_utils`. I moved most of the functionality into `TyCtxt`. I realized that the `librustc_trans_utils::common.rs` was just copy pasted from `librustc_trans::common.rs`.

Should I also get rid of the `librustc_trans::common.rs` in this PR? Most of the functionality seems a bit useless, I decided to put some of it into `TyCtxt` but maybe that is not the correct action here.

Should I also get rid of `librustc_trans_utils` completely here? Or should I do it in a separate PR?

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pnkfelix (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

eddyb commented:
r? @eddyb

eddyb commented on src/librustc/ty/instance.rs:
```diff
@@ -45,6 +45,51 @@ pub enum InstanceDef<'tcx> {
 }
 
+impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
+    pub fn is_inline_instance(
```
Where is this used? In general, I'd expect these methods to be on `Instance`, or where possible, `InstanceDef`, and private if only used locally. (I was trying to come up with better name for them)

MaikKlein replied:
This is only used in `requests_inline`afaik. But the original function is used in callee.rs and trans_item_rs. I can move it to instance, I wasn't sure where to put it because it takes an instance and a tyctxt.

eddyb commented on src/librustc/ty/instance.rs:
```diff
@@ -45,6 +45,51 @@ pub enum InstanceDef<'tcx> {
+        }
+    }
+
+    pub fn requests_inline(
```
I think this should be called `requires_local_instance` or something else along those lines.

eddyb commented on src/librustc/ty/mod.rs:
```diff
@@ -2380,6 +2380,21 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
 }
 
+impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
+    pub fn is_sized(self, ty: Ty<'tcx>) -> bool {
```
Ideally these two methods would be avoided. I wonder what the `DUMMY_SP` is used for - `is_sized` should never cause an ICE or an user error.

MaikKlein replied:
I wasn't sure about those functions, there are still a bunch of them in `rustc_trans::common.rs`. Initially I thought I would move them into `TyCtxt` but then I asked myself if they should even exist. Should I remove them?

eddyb replied:
Depends on what the usage looks like.

MaikKlein replied:
I haven't investigated it so far. These functions were just copied (not moved) into `trans_utils::common.rs` and therefore are only used in the collector IIRC. But the original functions are probably more widespread.

eddyb replied:
Replacing the uses in the collector with the more common forms is probably the best first step.

eddyb commented on src/librustc_mir/monomorphize/collector.rs:
```diff
@@ -25,21 +25,7 @@
-//!
-//!
-//! General Algorithm
+//! - Methm
```
This looks to be accidentally truncated.

MaikKlein replied:
Not sure what happened, I am fixing it now.

eddyb commented on src/librustc_trans/Cargo.toml:
```diff
@@ -27,6 +27,7 @@ rustc_incremental = { path = "../librustc_incremental" }
 rustc_llvm = { path = "../librustc_llvm" }
 rustc_platform_intrinsics = { path = "../librustc_platform_intrinsics" }
 rustc_trans_utils = { path = "../librustc_trans_utils" }
```
What's left of `rustc_trans_utils`?

MaikKlein replied:
https://github.com/MaikKlein/rust/tree/collector/src/librustc_trans_utils I am not sure where I should move the rest.

eddyb replied:
Oh I think you can throw them into `librustc/middle` if you want to move them. But nothing I see there should retain its current form anyway.

eddyb commented on src/librustc/ty/instance.rs:
```diff
@@ -45,18 +45,18 @@ pub enum InstanceDef<'tcx> {
     ) -> bool {
         use hir::map::DefPathData;
-        let def_id = match instance.def {
+        let def_id = match self.def {
```
With only `self.def` being accessed, doesn't this method belong on `InstanceDef`? Same for `requires_local_instance`.

MaikKlein replied:
You are right, I haven't looked to closely.

eddyb commented on src/librustc/ty/instance.rs:
```diff
@@ -45,6 +45,16 @@ pub enum InstanceDef<'tcx> {
 }
 
+impl<'a, 'tcx> Instance<'tcx> {
+    pub fn instance_ty(&self,
```
I'd call this `ty`. One thing I'm worried about is `trans_apply_param_substs` failing if this is accidentally used somewhere where a `ParamEnv` is needed or some such.
Shouldn't `self.def.def_ty(tcx).subst(tcx, self.substs)` suffice?

MaikKlein replied:
@eddyb I don't really know the details of substitutions yet. I have updated this function to use `self.def.def_ty(tcx).subst(tcx, self.substs)` but then I get an ICE in `run_cargo` (see latest build error). Everything seems to work when I use `trans_apply_param_substs`. 

Should I still avoid `trans_apply_param_substs`? I can investigate further why it crashes if you want me to.

eddyb commented on src/librustc/ty/mod.rs:
```diff
@@ -2380,6 +2380,17 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
 }
 
+impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
+    /// Given a DefId and some Substs, produces the monomorphic item type.
```
This shouldn't exist - maybe inline the body or use `Instance::new(def_id, substs).ty(tcx)` instead?

MaikKlein replied:
Oh sorry I should have looked closer, I am removing it now.

bors commented:
:umbrella: The latest upstream changes (presumably #44295) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #45554) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented on src/librustc_trans/debuginfo/metadata.rs:
```diff
@@ -1818,7 +1819,7 @@ pub fn create_global_var_metadata(cx: &CrateContext,
 
     let is_local_to_unit = is_node_local_to_unit(cx, node_id);
-    let variable_type = common::def_ty(cx.tcx(), node_def_id, Substs::empty());
+    let variable_type = Instance::new(node_def_id, Substs::empty()).ty(cx.tcx());
```
Oh wow. This should be just `cx.tcx().type_of(node_def_id)`.

eddyb commented on src/librustc_trans/debuginfo/mod.rs:
```diff
@@ -428,7 +428,7 @@ pub fn create_function_debug_context<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
             if cx.tcx().trait_id_of_impl(impl_def_id).is_none() {
                 let impl_self_ty =
-                    common::def_ty(cx.tcx(), impl_def_id, instance.substs);
+                    Instance::new(impl_def_id, instance.substs).ty(cx.tcx());
```
This shouldn't have been using `def_ty` at all... But `cx.tcx().type_of(impl_def_id).subst(cx.tcx(), instance.substs)`.

MaikKlein replied:
But that should be the same as `Instance::new(impl_def_id, instance.substs).ty(cx.tcx());`?

MaikKlein replied:
The error comes from https://github.com/MaikKlein/rust/blob/collector/src/librustc_trans/trans_item.rs#L174

MaikKlein replied:
Wait sorry it actually comes from `type_of`,

```
  10: rustc_trans::type_of::in_memory_type_of
  11: rustc_trans::trans_item::predefine_static
  12: rustc_trans::trans_item::MonoItemExt::predefine
  13: rustc_trans::base::compile_codegen_unit
  14: rustc::ty::maps::<impl rustc::ty::maps::queries::compile_codegen_unit<'tcx>>::compute_result
  15: rustc::dep_graph::graph::DepGraph::with_task_impl
  16: rustc::dep_graph::graph::DepGraph::with_task
  17: rustc_errors::Handler::track_diagnostics
  18: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  19: rustc::ty::maps::<impl rustc::ty::maps::queries::compile_codegen_unit<'tcx>>::force
  20: rustc::ty::maps::<impl rustc::ty::maps::queries::compile_codegen_unit<'tcx>>::try_get
  21: rustc::ty::maps::TyCtxtAt::compile_codegen_unit
  22: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::compile_codegen_unit
  23: rustc_trans::base::trans_crate
```

eddyb replied:
Does making *this* specific change cause an error? It's unclear to me, you seem to have replied pointing to a different change.

eddyb commented on src/librustc_trans/partitioning.rs:
```diff
@@ -587,7 +587,7 @@ fn characteristic_def_id_of_trans_item<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                 // This is a method within an inherent impl, find out what the
                 // self-type is:
-                let impl_self_ty = common::def_ty(tcx, impl_def_id, instance.substs);
+                let impl_self_ty = Instance::new(impl_def_id, instance.substs).ty(tcx);
```
Same here as https://github.com/rust-lang/rust/pull/45525#discussion_r147662872.

eddyb replied:
See https://github.com/rust-lang/rust/pull/45525#discussion_r152757123, still seems relevant to me (it's not a function or a global, but rather an `impl`'s `Self` type).

MaikKlein commented on src/librustc_trans/trans_item.rs:
```diff
@@ -172,7 +171,7 @@ fn predefine_static<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
     let def_id = ccx.tcx().hir.local_def_id(node_id);
     let instance = Instance::mono(ccx.tcx(), def_id);
-    let ty = common::instance_ty(ccx.tcx(), &instance);
+    let ty = instance.ty(ccx.tcx());
```
This line causes the error, I think.

MaikKlein commented:
@eddyb 

https://github.com/rust-lang/rust/pull/45525/files/d13f02127b00164de390c3fac36800dd09a0dd23..1c2abb2c01f5ac0ae6b64f735ea52290096dae29#diff-315d84e8fb6da1eb0de18431f7f7cbec
```
DEBUG:rustc_trans::type_of: type_of &mut str::Chars
DEBUG:rustc_trans::type_of: --> mapped t=&mut str::Chars to llty=%"str::Chars"*
DEBUG:rustc_trans::type_of: type_of &mut str::Bytes
DEBUG:rustc_trans::type_of: type_of str::Bytes
DEBUG:rustc_trans::type_of: --> mapped t=str::Bytes to llty=%"str::Bytes" = type opaque
DEBUG:rustc_trans::type_of: --> mapped t=&mut str::Bytes to llty=%"str::Bytes"*
DEBUG:rustc_trans::type_of: type_of result::Result<(), str::Utf8Error>
DEBUG:rustc_trans::type_of: --> mapped t=result::Result<(), str::Utf8Error> to llty=%"result::Result<(), str::Utf8Error>" = type { i64, [0 x i64], [2 x i64] }
DEBUG:rustc_trans::type_of: type_of [u8; <unevaluated[]>]
error: internal compiler error: unexpected panic
```
https://github.com/MaikKlein/rust/blob/collector/src/librustc_trans/type_of.rs#L152 I think this line will fail because the dimension of the array is `<unevaluated[]>`. How do I evaluate it without using `trans_apply_param_substs`?

bors commented:
:umbrella: The latest upstream changes (presumably #45247) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #45575) made this pull request unmergeable. Please resolve the merge conflicts.

carols10cents commented:
> How do I evaluate it without using trans_apply_param_substs?

@eddyb do you have an answer for [this question](https://github.com/rust-lang/rust/pull/45525#issuecomment-340541694) of @MaikKlein's?

eddyb commented:
@carols10cents We're actually waiting on #45225.

bors commented:
:umbrella: The latest upstream changes (presumably #45225) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
Time to get this rolling again!

MaikKlein commented:
It currently panics in `rustc_trans::debuginfo::metadata::fixed_vec_metadata`. I probably made a mistake when I rebased. I'll have a closer look at it tomorrow.

bors commented:
:umbrella: The latest upstream changes (presumably #45879) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented on src/librustc/ty/instance.rs:
```diff
@@ -68,6 +78,42 @@ impl<'tcx> InstanceDef<'tcx> {
+    ) -> bool {
+        use hir::map::DefPathData;
+        let def_id = match self {
+            &ty::InstanceDef::Item(def_id) => def_id,
```
This match should be on `*self`.

eddyb commented on src/librustc/ty/instance.rs:
```diff
@@ -68,6 +78,42 @@ impl<'tcx> InstanceDef<'tcx> {
+        if self.is_inline(tcx) {
+            return true
+        }
+        if let &ty::InstanceDef::DropGlue(..) = self {
```
Same here, `= *self`.

eddyb commented on src/librustc_mir/monomorphize/collector.rs:
```diff
@@ -790,7 +789,7 @@ fn find_vtable_types_for_unsizing<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                                             -> (Ty<'tcx>, Ty<'tcx>) {
     let ptr_vtable = |inner_source: Ty<'tcx>, inner_target: Ty<'tcx>| {
-        if type_has_metadata(tcx, inner_source) {
+        if inner_source.has_metadata(tcx) {
```
This check shouldn't be needed, the `else` branch does the right thing in all cases *anyway*.

eddyb commented on src/librustc/ty/mod.rs:
```diff
@@ -2009,6 +2009,20 @@ impl<'tcx> TyS<'tcx> {
         }
     }
+
+    pub fn has_metadata<'a>(&'tcx self, tcx: TyCtxt<'a, 'tcx, 'tcx>) -> bool {
```
The way this is written isn't very "portable", so I'd rather this be moved back to `rustc_trans` (in the only place where it's used).

eddyb commented on src/librustc_trans/partitioning.rs:
```diff
@@ -102,21 +102,21 @@
 
-use collector::InliningMap;
-use common;
+use rustc_mir::monomorphize::collector::InliningMap;
```
Hmmm, doesn't this entire file (`partitioning.rs`) belong in `rustc_mir/monomorphize`?

MaikKlein commented:
@eddyb I addressed a few of your comments in those PR.

Regarding the irc comment 

>  to expand on my original point, I didn't want *more* than one trans_apply_param_substs call in total, if Instance::ty was replaced in some cases with a more direct tcx.type_of or w/e, it should work *without* trans_apply_param_substs, but Instance::ty itself still needs it (possibly not for long though, it's just unclear how many things would have to be fixed if it didn't do trans_apply_param_substs)

I only touched two cases that you outlined in the comments so far. `Instance::ty` still uses `trans_apply_param_substs` currently :(. Do you want to me remove that `trans_apply_param_substs` from `Instance::ty` again? I could then use trial and error to see where `trans_apply_param_substs` really is needed.

I also moved paritioning.rs to rustc_mir. Reexporting got a bit confusing, which I need to cleanup.

bors commented:
:umbrella: The latest upstream changes (presumably #46370) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #46509) made this pull request unmergeable. Please resolve the merge conflicts.

carols10cents commented:
@eddyb review ping for you!

bors commented:
:umbrella: The latest upstream changes (presumably #45002) made this pull request unmergeable. Please resolve the merge conflicts.

MaikKlein commented:
@eddyb r?

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit ba194eb has been approved by `eddyb`

<!-- @bors r=eddyb ba194eb3bf28ae32c51abf0e487e6359b8450acb -->

bors commented:
:hourglass: Testing commit ba194eb3bf28ae32c51abf0e487e6359b8450acb with merge 1686d719899821832d2fda6680fb8b6510ee43fc...

bors commented:
:umbrella: The latest upstream changes (presumably #45047) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 2e49674 has been approved by `eddyb`

<!-- @bors r=eddyb 2e49674e52729eef3bb66fbd0b4baaa2ea7bf6ef -->

bors commented:
:lock: Merge conflict

arielb1 commented:
@bors r=eddyb

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #46701

bors commented:
:pushpin: Commit 2e49674 has been approved by `eddyb`

<!-- @bors r=eddyb 2e49674e52729eef3bb66fbd0b4baaa2ea7bf6ef -->

bors commented:
:lock: Merge conflict

nagisa commented:
@bors r=eddyb

bors commented:
:pushpin: Commit 1051634 has been approved by `eddyb`

<!-- @bors r=eddyb 1051634610d0778b2f25a9e8f8277eca8c8088b5 -->

bors commented:
:lock: Merge conflict

arielb1 commented:
@bors p=10

Let's avoid rollups backing this out

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 7fe5398 has been approved by `arielb1`

<!-- @bors r=arielb1 7fe539804898ebbcb1d2287375295e72420c9571 -->

bors commented:
:lock: Merge conflict

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 9b60bdc has been approved by `arielb1`

<!-- @bors r=arielb1 9b60bdcc556a765d01b5650720b9f82916f6262b -->

bors commented:
:hourglass: Testing commit 9b60bdcc556a765d01b5650720b9f82916f6262b with merge badf9f4fb5c7e34c4111b321613b2294fa3e28ab...

arielb1 commented:
@bors r=eddyb

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request is currently being tested. If there's no response from the continuous integration service, you may use `retry` to trigger a build again.

bors commented:
:pushpin: Commit 9b60bdc has been approved by `eddyb`

<!-- @bors r=eddyb 9b60bdcc556a765d01b5650720b9f82916f6262b -->

bors commented:
:hourglass: Testing commit 9b60bdcc556a765d01b5650720b9f82916f6262b with merge fcbbe4991a77542afc398319db9eb9975f745217...

arielb1 commented:
@bors retry

bors commented:
:hourglass: Testing commit 9b60bdcc556a765d01b5650720b9f82916f6262b with merge f808be2b6aacde7205b35b4bccdba1631ae048cf...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/317136689?utm_source=github_status&utm_medium=notification)

est31 commented:
Error while building stage1 std:
```
[00:20:44] error: internal compiler error: /checkout/src/librustc_trans/debuginfo/type_names.rs:177: debuginfo: Trying to create type name for unexpected type: <iter::AlwaysOk<usize> as ops::try::Try>::Ok
[00:20:44] 
[00:20:44] note: the compiler unexpectedly panicked. this is a bug.
```

Looks like a legitimate failure @MaikKlein

MaikKlein commented:
I am a little confused. All the tests pass on my local machine and they also passed on travis before.

When bors runs `Testing commit 9b60bdc with merge f808be2...` is that different from the CI that runs when I push a new commit?

Luckily I only had a few changes and I think I know where the problem is.

est31 commented:
> When bors runs Testing commit 9b60bdc with merge f808be2... is that different from the CI that runs when I push a new commit?

Both the tests being run are different, as well as the commit used to run the tests.

The bors tests are more extensive, however the travis test includes a full boostrap as well, and it was failing inside bootstrap. So idk. Either the ICE doesn't occur on the special platform that travis has, or it is due to the differences in the commit.

The commit that bors uses is different from the commit of your PR. Bors first merges your PR into master and then runs tests, while travis takes the bare commit of your PR. So bors has all the changes that have happened on master since you last rebased, while those changes lack for travis.

Can you reproduce the issue after you've rebased to latest master?

MaikKlein commented:
>Can you reproduce the issue after you've rebased to latest master?

Everything still succeeds on my local machine but I can run those tests with `./src/ci/docker/run.sh dist-x86_64-freebsd` which shows me a different panic. I should now be able to fix it :).

arielb1 commented:
@MaikKlein 

I can reproduce this ICE locally and am investigating it

eddyb commented on src/librustc_mir/interpret/terminator/mod.rs:
```diff
@@ -73,8 +73,7 @@ impl<'a, 'tcx, M: Machine<'tcx>> EvalContext<'a, 'tcx, M> {
                         let fn_ptr = self.value_to_primval(func)?.to_ptr()?;
                         let instance = self.memory.get_fn(fn_ptr)?;
                         // FIXME(eddyb) use `Instance::ty` when it becomes available.
```
You can remove the comment.

eddyb commented on src/librustc_mir/interpret/step.rs:
```diff
@@ -173,8 +173,7 @@ impl<'a, 'tcx, M: Machine<'tcx>> EvalContext<'a, 'tcx, M> {
             return Ok(false);
         }
         // FIXME(eddyb) use `Instance::ty` when it becomes available.
```
You can remove the comment.

eddyb commented on src/librustc_mir/interpret/const_eval.rs:
```diff
@@ -38,8 +38,7 @@ pub fn eval_body<'a, 'tcx>(
             return Err(ConstEvalError::NotConst("extern global".to_string()).into());
         }
         // FIXME(eddyb) use `Instance::ty` when it becomes available.
```
You can remove the comment.

eddyb commented on src/librustc/traits/trans/mod.rs:
```diff
@@ -119,6 +120,12 @@ impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
+    pub fn trans_impl_self_ty(&self, def_id: DefId, substs: &'tcx Substs<'tcx>)
+                              -> Ty<'tcx>
+    {
+        self.trans_apply_param_substs(substs, &self.type_of(def_id))
```
This makes me sad but I guess I'm okay with it for now if it's necessary.

arielb1 replied:
It just feels like less of a footgun than the other option.

arielb1 commented:
NOTE: miri is currently broken - even without this PR - because of https://github.com/solson/miri/pull/359/files#r157357606 - cc @oli-obk

arielb1 commented:
r? @eddyb for the last commit

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit fcd2d06 has been approved by `eddyb`

<!-- @bors r=eddyb fcd2d06890feb94f330120af64eedbf7b9a07f33 -->

bors commented:
:hourglass: Testing commit fcd2d06890feb94f330120af64eedbf7b9a07f33 with merge 9d6086c185caa6f38a3556a766257cdb7bdf0cab...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/317595330?utm_source=github_status&utm_medium=notification)

kennytm commented:
solson/miri@a23e587dc3a55a766ddd95454fd8368988cc7a3a cannot be found in the solson/miri repository. 

solson/miri#359 must be placed into a branch of the solson/miri repository for this to pass (or keep using the master branch of miri, and just break the tool).

arielb1 commented:
I suppose I'll break miri then, given that I already have the fix in a PR.

@bors r+

bors commented:
:pushpin: Commit 91dbc7d has been approved by `arielb1`

<!-- @bors r=arielb1 91dbc7dab0cf07806d2c53c5ed4ef87e7b744d07 -->

arielb1 commented:
@bors r=eddyb

bors commented:
:pushpin: Commit effd99f has been approved by `eddyb`

<!-- @bors r=eddyb effd99fd81dbf83c15519d6dd4fd6bb4bcbd6396 -->

arielb1 commented:
@bors r=eddyb

bors commented:
:pushpin: Commit b6a7688 has been approved by `eddyb`

<!-- @bors r=eddyb b6a768861d44c0c24be3156713cdbe04428406fd -->

bors commented:
:hourglass: Testing commit b6a768861d44c0c24be3156713cdbe04428406fd with merge b36b12eed2daecdf436755834f9e858141d212eb...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/318172212?utm_source=github_status&utm_medium=notification)

eddyb commented:
Need to undo the miri submodule change.

arielb1 commented:
@bors r=eddyb

bors commented:
:pushpin: Commit 6e78b66 has been approved by `eddyb`

<!-- @bors r=eddyb 6e78b665786b3f37730d5af3363ce74ad832282d -->

bors commented:
:hourglass: Testing commit 6e78b665786b3f37730d5af3363ce74ad832282d with merge 0a440e76bff84933783d590cf47ee811c5b345a0...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/318251343?utm_source=github_status&utm_medium=notification)

kennytm commented:
@bors retry #46313.

bors commented:
:hourglass: Testing commit 6e78b665786b3f37730d5af3363ce74ad832282d with merge 68e6abacfa269e101e7f6f2e7f83d0527012ae0a...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5748)

bors commented:
:hourglass: Testing commit 6e78b665786b3f37730d5af3363ce74ad832282d with merge b76f224af867b198e4051806cc6e1ac686085cdc...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5749), [status-travis](https://travis-ci.org/rust-lang/rust/builds/318396486?utm_source=github_status&utm_medium=notification)
Approved by: eddyb
Pushing b76f224af867b198e4051806cc6e1ac686085cdc to master...
